File: sling/nlp/document/lexicon.cc
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "sling/nlp/document/lexicon.h"

#include "sling/base/types.h"
#include "sling/nlp/document/affix.h"
#include "sling/stream/memory.h"
#include "sling/util/vocabulary.h"

namespace sling {
namespace nlp {

void WordShape::Extract(const string &word) {
  quote = NO_QUOTE;
  hyphen = NO_HYPHEN;
  bool has_upper = false;
  bool has_lower = false;
  bool has_punctuation = false;
  bool all_punctuation = !word.empty();
  bool has_digit = false;
  bool all_digit = !word.empty();
  const char *p = word.data();
  const char *end = p + word.size();
  while (p < end) {
    int code = UTF8::Decode(p);
    int cat = Unicode::Category(code);

    // Hyphenation.
    if (cat == CHARCAT_DASH_PUNCTUATION) hyphen = HAS_HYPHEN;

    // Capitalization.
    if (Unicode::IsUpper(code)) has_upper = true;
    if (Unicode::IsLower(code)) has_lower = true;

    // Punctuation.
    bool is_punct = Unicode::IsPunctuation(code);
    all_punctuation &= is_punct;
    has_punctuation |= is_punct;

    // Quotes.
    switch (cat) {
      case CHARCAT_INITIAL_QUOTE_PUNCTUATION:
        quote = OPEN_QUOTE;
        break;
      case CHARCAT_FINAL_QUOTE_PUNCTUATION:
        quote = CLOSE_QUOTE;
        break;
      case CHARCAT_OTHER_PUNCTUATION:
        if (code == '\'' || code == '"') quote = UNKNOWN_QUOTE;
        break;
      case CHARCAT_MODIFIER_SYMBOL:
        if (code == '`') quote = UNKNOWN_QUOTE;
        break;
    }

    // Digits.
    bool is_digit = Unicode::IsDigit(code);
    all_digit &= is_digit;
    has_digit |= is_digit;

    p = UTF8::Next(p);
  }

  // Compute word capitalization.
  if (!has_upper && has_lower) {
    capitalization = LOWERCASE;
  } else if (has_upper && !has_lower) {
    capitalization = UPPERCASE;
  } else if (!has_upper && !has_lower) {
    capitalization = NON_ALPHABETIC;
  } else {
    capitalization = CAPITALIZED;
  }

  // Compute punctuation feature.
  if (all_punctuation) {
    punctuation = ALL_PUNCTUATION;
  } else if (has_punctuation) {
    punctuation = SOME_PUNCTUATION;
  } else {
    punctuation = NO_PUNCTUATION;
  }

  // Compute quote feature.
  if (quote != NO_QUOTE) {
    // Penn Treebank open and close quotes are multi-character.
    if (word == "``") quote = OPEN_QUOTE;
    if (word == "''") quote = CLOSE_QUOTE;
  }

  // Compute digit feature.
  if (all_digit) {
    digit = ALL_DIGIT;
  } else if (has_digit) {
    digit = SOME_DIGIT;
  } else {
    digit = NO_DIGIT;
  }
}

void Lexicon::InitWords(Vocabulary::Iterator *words) {
  // Initialize mapping from words to ids.
  vocabulary_.Init(words);

  // Initialize mapping from ids to words.
  words_.resize(words->Size());
  int index = 0;
  words->Reset();
  Text word;
  while (words->Next(&word, nullptr)) {
    words_[index].word.assign(word.data(), word.size());
    index++;
  }
}

void Lexicon::WriteVocabulary(string *buffer, char terminator) const {
  for (const Entry &e : words_) {
    buffer->append(e.word);
    buffer->push_back(terminator);
  }
}

void Lexicon::InitPrefixes(const char *data, size_t size) {
  // Read prefixes.
  ArrayInputStream stream(data, size);
  prefixes_.Read(&stream);

  // Pre-compute the longest prefix for all words in lexicon.
  for (Entry &entry : words_) {
    entry.prefix = prefixes_.GetLongestAffix(entry.word);
  }
}

void Lexicon::InitSuffixes(const char *data, size_t size) {
  // Read suffixes.
  ArrayInputStream stream(data, size);
  suffixes_.Read(&stream);

  // Pre-compute the longest suffix for all words in lexicon.
  for (Entry &entry : words_) {
    entry.suffix = suffixes_.GetLongestAffix(entry.word);
  }
}

void Lexicon::WritePrefixes(string *buffer) const {
  StringOutputStream stream(buffer);
  prefixes_.Write(&stream);
}

void Lexicon::WriteSuffixes(string *buffer) const {
  StringOutputStream stream(buffer);
  suffixes_.Write(&stream);
}

void Lexicon::BuildPrefixes(int max_prefix) {
  prefixes_.Reset(max_prefix);
  prefixes_.AddAffixesForWord("");
  if (max_prefix > 0) {
    for (Entry &entry : words_) {
      entry.prefix = prefixes_.AddAffixesForWord(entry.word);
    }
  }
}

void Lexicon::BuildSuffixes(int max_suffix) {
  suffixes_.Reset(max_suffix);
  suffixes_.AddAffixesForWord("");
  if (max_suffix > 0) {
    for (Entry &entry : words_) {
      entry.suffix = suffixes_.AddAffixesForWord(entry.word);
    }
  }
}

void Lexicon::PrecomputeShapes() {
  for (Entry &entry : words_) {
    entry.shape.Extract(entry.word);
  }
}

int Lexicon::Lookup(const string &word,
                    Affix **prefix, Affix **suffix,
                    WordShape *shape) const {
  // Normalize word.
  string normalized;
  UTF8::Normalize(word, normalization_, &normalized);

  // Look up word in vocabulary.
  int id = vocabulary_.Lookup(normalized);

  // Return pre-computed information from the lexicon for known words.
  if (id != -1) {
    const Entry &entry = words_[id];
    *prefix = entry.prefix;
    *suffix = entry.suffix;
    *shape = entry.shape;

    // Due to normalization of the words in the lexicon, some of the
    // pre-computed features need to be re-computed.
    if (normalization_ & NORMALIZE_CASE) {
      // Re-compute capitalization for case-normalized lexicon.
      bool has_upper = false;
      bool has_lower = false;
      const char *p = word.data();
      const char *end = p + word.size();
      while (p < end) {
        int code = UTF8::Decode(p);
        if (Unicode::IsUpper(code)) has_upper = true;
        if (Unicode::IsLower(code)) has_lower = true;
        p = UTF8::Next(p);
      }

      if (!has_upper && has_lower) {
        shape->capitalization = WordShape::LOWERCASE;
      } else if (has_upper && !has_lower) {
        shape->capitalization = WordShape::UPPERCASE;
      } else if (!has_upper && !has_lower) {
        shape->capitalization = WordShape::NON_ALPHABETIC;
      } else {
        shape->capitalization = WordShape::CAPITALIZED;
      }
    }
    if (normalization_ & NORMALIZE_PUNCTUATION) {
      // Re-compute punctuation for punctuation-normalized lexicon.
      bool has_punctuation = false;
      bool all_punctuation = !word.empty();
      const char *p = word.data();
      const char *end = p + word.size();
      while (p < end) {
        int code = UTF8::Decode(p);
        bool is_punct = Unicode::IsPunctuation(code);
        all_punctuation &= is_punct;
        has_punctuation |= is_punct;
        p = UTF8::Next(p);
      }

      if (all_punctuation) {
        shape->punctuation = WordShape::ALL_PUNCTUATION;
      } else if (has_punctuation) {
        shape->punctuation = WordShape::SOME_PUNCTUATION;
      } else {
        shape->punctuation = WordShape::NO_PUNCTUATION;
      }
    }

    return id;
  }

  // Compute affixes and shape features on-the-fly for unknown words.
  if (prefixes_.max_length() > 0) {
    *prefix = prefixes_.GetLongestAffix(normalized);
  } else {
    *prefix = nullptr;
  }
  if (suffixes_.max_length() > 0) {
    *suffix = suffixes_.GetLongestAffix(normalized);
  } else {
    *suffix = nullptr;
  }
  shape->Extract(word);
  return oov_;
}

int Lexicon::Lookup(const string &word) const {
  // Look up word in vocabulary.
  int id = vocabulary_.Lookup(word);
  return id == -1 ? oov_ : id;
}

}  // namespace nlp
}  // namespace sling

File: sling/nlp/parser/trainer/lexicon.py
# Copyright 2017 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http:#www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


# Class for computing and serving a lexicon.
# Usage:
#   lexicon = Lexicon(normalize_digits=...)
#   lexicon.add("foo")
#   lexicon.add("bar")
#   ...
#   index = lexicon.index("foo")
#
# Methods with names starting with an underscore are meant to be internal.
class Lexicon:
  def __init__(self, normalize_digits=True, oov_item="<UNKNOWN>"):
    self.normalize_digits = normalize_digits
    self.item_to_index = {}
    self.index_to_item = {}

    if oov_item is not None:
      self.oov_item = oov_item
      self.oov_index = 0   # Don't change this; OOV is always at position 0
      self.index_to_item[self.oov_index] = self.oov_item
      self.item_to_index[self.oov_item] = self.oov_index
    else:
      self.oov_item = None
      self.oov_index = None


  # Returns whether the lexicon has an OOV item.
  def has_oov(self):
    return self.oov_index is not None


  # Returns the key internally used by the lexicon.
  def _key(self, item):
    if self.normalize_digits:
      return "".join([c if not c.isdigit() else '9' for c in list(item)])
    else:
      return item


  # Loads the lexicon from a text file with one key per line.
  def load(self, vocabfile):
    with open(vocabfile, "r") as f:
      index = 0
      for line in f:
        line = line.strip()
        if line == self.oov_item:
          assert index == self.oov_index, index
        self.item_to_index[line] = index
        self.index_to_item[index] = line
        index += 1


  # Reads the lexicon from a delimited string.
  def read(self, vocabstring, delimiter):
    index = 0
    lines = vocabstring.split(delimiter)
    if lines[-1] == '': lines.pop()
    for line_index, line in enumerate(lines):
      if line == self.oov_item:
        assert index == self.oov_index, index
      self.item_to_index[line] = index
      self.index_to_item[index] = line
      index += 1


  # Adds 'item' to the lexicon.
  def add(self, item):
    item = self._key(item)
    if item not in self.item_to_index:
      i = len(self.item_to_index)
      self.item_to_index[item] = i
      self.index_to_item[i] = item


  # Returns the size of the lexicon, including the OOV item, if any.
  def size(self):
    return len(self.item_to_index)


  # Returns the integer index of 'item' in the lexicon, if present.
  def index(self, item):
    item = self._key(item)
    if item not in self.item_to_index:
      return self.oov_index  # this is None if !has_oov()
    else:
      return self.item_to_index[item]


  # Returns a string representation of the key whose id is 'index'.
  def value(self, index):
    assert index >= 0 and index < len(self.index_to_item), "%r" % index
    return self.index_to_item[index]


  # Returns a string representation of the lexicon.
  def __str__(self):
    s = [self.index_to_item[i] for i in range(self.size())]
    return "\n".join(s)


  # Returns the string representation of the first 'n' keys in the lexicon.
  def first_few(self, prefix="", n=100):
    s = []
    for i in range(min(n, self.size())):
      s.append(prefix + str(i) + " = " + self.index_to_item[i])
    return "\n".join(s)
